src/AggregateFunctions/AggregateFunctionGroupConcat.cpp

```diff
@@ -0,0 +1,297 @@
+    {
+        auto & cur_data = this->data(place);
+
+        readVarUInt(cur_data.size, buf);
+        readVarUInt(cur_data.allocated_size, buf);
+
+        readBinary(cur_data.is_empty, buf);
+
+        buf.readStrict(cur_data.data, cur_data.size);
+
+        if constexpr (has_limit)
+        {
+            readVarUInt(cur_data.count, buf);
+            cur_data.offsets.resize_exact(cur_data.count * 2, arena);
+            for (auto & offset : cur_data.offsets)
+                readVarUInt(offset, buf);
+        }
+    }
+
+    void insertResultInto(AggregateDataPtr __restrict place, IColumn & to, Arena *) const override
+    {
+        auto & cur_data = this->data(place);
+
+        if (cur_data.is_empty)
+        {
+            auto & column_string = assert_cast<ColumnString &>(to);
+            column_string.insert(NULL);
+            return;
+        }
+
+        auto & column_string = assert_cast<ColumnString &>(to);
+        column_string.insertData(cur_data.data, cur_data.size);
+    }
+
+    bool allocatesMemoryInArena() const override { return true; }
+};
+
+AggregateFunctionPtr createAggregateFunctionGroupConcat(
+    const std::string & name, const DataTypes & argument_types, const Array & parameters, const Settings *)
+{
+    assertUnary(name, argument_types);
+
+    bool has_limit = false;
+    UInt64 max_elems = 0;
+    String delimiter;
+
+    if (parameters.size() > 2)
+        throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+            "Incorrect number of parameters for aggregate function {}, should be 0, 1 or 2", name);
+
+    if (!parameters.empty())
+    {
+        auto type = parameters[0].getType();
+        if (type != Field::Types::String)
+               throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "First parameter for aggregate function {} should be string", name);
+
+        delimiter = parameters[0].get<String>();
+    }
+    if (parameters.size() == 2)
+    {
+        auto type = parameters[1].getType();
+        if (type != Field::Types::Int64 && type != Field::Types::UInt64)
+               throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Second parameter for aggregate function {} should be positive number", name);
+
+        if ((type == Field::Types::Int64 && parameters[1].get<Int64>() <= 0) ||
+            (type == Field::Types::UInt64 && parameters[1].get<UInt64>() == 0))
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Second parameter for aggregate function {} should be positive number", name);
+
+        has_limit = true;
+        max_elems = parameters[1].get<UInt64>();
+    }
+
+    if (has_limit)
+        return std::make_shared<GroupConcatImpl</* has_limit= */ true>>(argument_types[0], parameters, max_elems, delimiter);
+    else
+        return std::make_shared<GroupConcatImpl</* has_limit= */ false>>(argument_types[0], parameters, max_elems, delimiter);
```

Comment:
I hope we don't need to put max_elems into GroupConcatImpl if we don't have a limit, otherwise, we can make only one line of code instead of 4: 

return std::make_shared<GroupConcatImpl</* has_limit= */ has_limit>>(argument_types[0], parameters, max_elems, delimiter);

Reply:
I made 4 lines because template argument should be compile-time constant and we can't put has_limit there